### ISSUE TITLE
Gateway timeout (504) when trying to access duplicate finder

### ISSUE THREAD

**Issue by remz1337** [eldertek/duplicatefinder#43]
Continuing our discussion from https://github.com/eldertek/duplicatefinder/issues/26#issuecomment-1868591504

Similar issue, when I run `occ duplicates:list` everything shows up quite fast, `occ duplicates:list --user USER` takes quite more time. When I try to open the web app, after a few seconds there are 2 error pop ups saying `Could not fetch (un)acknowledged duplicates`, then pages says `No duplicates found or no duplicate selected.`. The logs show that the server hit a timeout (log below).

If it's due to the time it takes to build the list, I see 2 potential solutions:
- Use pagination (show 50 per page, then the page needs to load only the first few and not the entire list)
- Some form of dynamic/async load

Web console log:

`GET https://nextcloud.example.com/apps/duplicatefinder/api/duplicates/acknowledged?page=1 504 (Gateway Timeout)`

```
{
    "message": "Request failed with status code 504",
    "name": "AxiosError",
    "stack": "AxiosError: Request failed with status code 504\n    at https://nextcloud.example.com/apps/duplicatefinder/js/duplicatefinder-main.js?v=368756b4-0:2:792583\n    at XMLHttpRequest.d (https://nextcloud.example.com/apps/duplicatefinder/js/duplicatefinder-main.js?v=368756b4-0:2:792728)",
    "config": {
        "transitional": {
            "silentJSONParsing": true,
            "forcedJSONParsing": true,
            "clarifyTimeoutError": false
        },
        "adapter": [
            "xhr",
            "http"
        ],
        "transformRequest": [
            null
        ],
        "transformResponse": [
            null
        ],
        "timeout": 0,
        "xsrfCookieName": "XSRF-TOKEN",
        "xsrfHeaderName": "X-XSRF-TOKEN",
        "maxContentLength": -1,
        "maxBodyLength": -1,
        "env": {},
        "headers": {
            "Accept": "application/json, text/plain, */*",
            "requesttoken": "8M8FeZIL6OsSRag7WvFmlHt5ew27ZiTBVuXpfqTmdiI=:p5SAStQ/qql4at8CEbch1hkoIkvALhT0LqDGB+3XRkU="
        },
        "method": "get",
        "url": "/apps/duplicatefinder/api/duplicates/acknowledged?page=1"
    },
    "code": "ERR_BAD_RESPONSE",
    "status": 504
}

```

**Comment by whitecloudy** [comment]
Did you check your DB works fine? And have you tried clear your duplicates information and re-scan it?
It is not exact solution but it works for me.

**Comment by eldertek** [comment]
> Continuing our discussion from [#26 (comment)](https://github.com/eldertek/duplicatefinder/issues/26#issuecomment-1868591504)
> 
> Similar issue, when I run `occ duplicates:list` everything shows up quite fast, `occ duplicates:list --user USER` takes quite more time. When I try to open the web app, after a few seconds there are 2 error pop ups saying `Could not fetch (un)acknowledged duplicates`, then pages says `No duplicates found or no duplicate selected.`. The logs show that the server hit a timeout (log below).
> 
> If it's due to the time it takes to build the list, I see 2 potential solutions:
> 
> * Use pagination (show 50 per page, then the page needs to load only the first few and not the entire list)
> * Some form of dynamic/async load
> 
> Web console log:
> 
> `GET https://nextcloud.example.com/apps/duplicatefinder/api/duplicates/acknowledged?page=1 504 (Gateway Timeout)`
> 
> ```
> {
>     "message": "Request failed with status code 504",
>     "name": "AxiosError",
>     "stack": "AxiosError: Request failed with status code 504\n    at https://nextcloud.example.com/apps/duplicatefinder/js/duplicatefinder-main.js?v=368756b4-0:2:792583\n    at XMLHttpRequest.d (https://nextcloud.example.com/apps/duplicatefinder/js/duplicatefinder-main.js?v=368756b4-0:2:792728)",
>     "config": {
>         "transitional": {
>             "silentJSONParsing": true,
>             "forcedJSONParsing": true,
>             "clarifyTimeoutError": false
>         },
>         "adapter": [
>             "xhr",
>             "http"
>         ],
>         "transformRequest": [
>             null
>         ],
>         "transformResponse": [
>             null
>         ],
>         "timeout": 0,
>         "xsrfCookieName": "XSRF-TOKEN",
>         "xsrfHeaderName": "X-XSRF-TOKEN",
>         "maxContentLength": -1,
>         "maxBodyLength": -1,
>         "env": {},
>         "headers": {
>             "Accept": "application/json, text/plain, */*",
>             "requesttoken": "8M8FeZIL6OsSRag7WvFmlHt5ew27ZiTBVuXpfqTmdiI=:p5SAStQ/qql4at8CEbch1hkoIkvALhT0LqDGB+3XRkU="
>         },
>         "method": "get",
>         "url": "/apps/duplicatefinder/api/duplicates/acknowledged?page=1"
>     },
>     "code": "ERR_BAD_RESPONSE",
>     "status": 504
> }
> ```

In fact it is already asynchronous

**Comment by remz1337** [comment]
> Did you check your DB works fine? And have you tried clear your duplicates information and re-scan it? It is not exact solution but it works for me.

I'm using PostgreSQL DB. I followed the link in your other comment but can't find the procedure you followed. Can you tell me which tables exactly I need to empty/drop in the DB?

**Comment by whitecloudy** [comment]
> > Did you check your DB works fine? And have you tried clear your duplicates information and re-scan it? It is not exact solution but it works for me.
> 
> @whitecloudy I'm using PostgreSQL DB. I followed the link in your other comment but can't find the procedure you followed. Can you tell me which tables exactly I need to empty/drop in the DB?
> 
> Edit: When you say clean+rescan, do you mean the 2 buttons in the admin page of DuplicateFinder?

The link that i posted was about MariaDB docker container's environment variable. So i'm not sure it works for PostgreSQL DB. If your DB works perfectly(what I mean perfect is DB does not put out any kind of error message), then you can just skip that part.

And yes, you can do clean and rescan by 2 buttons in the admin page. Or you can just use occ command.(what I did)

**Comment by remz1337** [comment]
Indeed, doing a `clean` + `rescan` did the trick. I let the rescan work overnight. Now, when I open the DuplicateFinder page in Nextcloud, it shows only ~10 duplicates, and after about a minute all the rest of the duplicates (I'm guessing something like >100k) starts showing.

And then it gets laggy and the scrollbar disappears. At least it is usable now.

I feel like this should still be investigated. Or at least write it in the docs/install instructions that one needs to run a full clean+rescan after install.

**Comment by eldertek** [comment]
> Indeed, doing a `clean` + `rescan` did the trick. I let the rescan work overnight. Now, when I open the DuplicateFinder page in Nextcloud, it shows only ~10 duplicates, and after about a minute all the rest of the duplicates (I'm guessing something like >100k) starts showing.
> 
> And then it gets laggy and the scrollbar disappears. At least it is usable now.
> 
> I feel like this should still be investigated. Or at least write it in the docs/install instructions that one needs to run a full clean+rescan after install.

Thank you for your experience, I will add something like in doc, or try to investigate

**Comment by eldertek** [comment]
Should be fixed by v1.1.5, only on nextcloud 28.
